server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -115,13 +113,14 @@ private DocumentMapper(IndexSettings indexSettings,
         this.documentParser = documentParser;
         this.indexSettings = indexSettings;
         this.indexAnalyzers = indexAnalyzers;
-        this.fieldMappers = MappingLookup.fromMapping(this.mapping);
 
         try {
             mappingSource = new CompressedXContent(this, XContentType.JSON, ToXContent.EMPTY_PARAMS);
         } catch (Exception e) {
             throw new ElasticsearchGenerationException("failed to serialize source for type [" + type + "]", e);
         }
+        // TODO this::parse performs a volatile read on mapping from MapperService. Yikes!
```

Comment:
Yeah. I was just realizing last week that we're ok here. I'll drop it. It is pretty tricky though!

Reply:
Agreed, I have some plans to maybe make it a bit better, but still, Mapping and some other stuff is needed all the way in ParseContext. I was wondering if parse does fit in MappingLookup, and I think the answer is that we don't have many alternatives. The only one I can think of is forget about percolator and let it do a volatile read, but it would be a shame in the context of this whole PR, so I think we are good.